### PR TITLE
Move where cuda implementation to TensorIterator

### PR DESCRIPTION
**Pull request by zasdfgbnm**
`where` is special because the arguments do not have the same type, which does not satisfy the assumption in modern https://github.com/pytorch/pytorch/pull/32383. I migrate it to TensorIterator so that there is something to test that this case is not broken. Currently, this case fallback to using legacy (not vectorized, not unrolled) code. It should be supported in the future when I cleanup `Loops.cuh`.

I also move some sharing part of `CUDALoops.cuh` and `ROCmLoops.cuh` into `Loops.cuh` so that to logic for checking whether `func_t` has the same arg types could be shared.

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit f8cb213b:


**None of the build failures appear to be your fault**.

* **2/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f8cb213b3de49110bd2131d8df4c947bd2263f9d).

### :snowflake: 2 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4413519) pytorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=95180683) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=95180683)) :snowflake:
<details>
<summary>
<code>Feb 04 20:48:54   Error generating file</code>
</summary>

```
Feb 04 20:48:54           detected during: 
Feb 04 20:48:54             instantiation of "void at::native::gpu_kernel_impl(at::TensorIterator &, const func_t &) [with func_t=lambda [](__nv_bool, scalar_t, scalar_t)->scalar_t]"  
Feb 04 20:48:54 (384): here 
Feb 04 20:48:54             instantiation of "void at::native::gpu_kernel(at::TensorIterator &, const func_t &) [with func_t=lambda [](__nv_bool, scalar_t, scalar_t)->scalar_t]"  
Feb 04 20:48:54 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/TensorCompare.cu(15): here 
Feb 04 20:48:54  
Feb 04 20:48:54 Error limit reached. 
Feb 04 20:48:54 100 errors detected in the compilation of "/tmp/tmpxft_000003d9_00000000-6_TensorCompare.cpp1.ii". 
Feb 04 20:48:54 Compilation terminated. 
Feb 04 20:48:54 CMake Error at torch_cuda_generated_TensorCompare.cu.o.Release.cmake:281 (message): 
Feb 04 20:48:54   Error generating file 
Feb 04 20:48:54   /var/lib/jenkins/workspace/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_TensorCompare.cu.o 
Feb 04 20:48:54  
Feb 04 20:48:54  
Feb 04 20:48:54 caffe2/CMakeFiles/torch_cuda.dir/build.make:917: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_TensorCompare.cu.o' failed 
Feb 04 20:48:54 make[2]: *** [caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_TensorCompare.cu.o] Error 1 
Feb 04 20:48:54 make[2]: *** Waiting for unfinished jobs.... 
Feb 04 20:48:54 /var/lib/jenkins/workspace/c10/util/ArrayRef.h(278): warning: attribute does not apply to any entity 
Feb 04 20:48:54  
Feb 04 20:48:54 /var/lib/jenkins/workspace/c10/util/ArrayRef.h(278): warning: attribute does not apply to any entity 
Feb 04 20:48:54  
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4413535) pytorch_libtorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=95181375) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=95181375)) :snowflake:
<details>
<summary>
<code>Feb 04 20:49:43   Error generating file</code>
</summary>

```
Feb 04 20:49:43 (384): here 
Feb 04 20:49:43             instantiation of "void at::native::gpu_kernel(at::TensorIterator &, const func_t &) [with func_t=lambda [](__nv_bool, scalar_t, scalar_t)->scalar_t]"  
Feb 04 20:49:43 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/TensorCompare.cu(15): here 
Feb 04 20:49:43  
Feb 04 20:49:43 Error limit reached. 
Feb 04 20:49:43 100 errors detected in the compilation of "/tmp/tmpxft_00007fbe_00000000-6_TensorCompare.cpp1.ii". 
Feb 04 20:49:43 Compilation terminated. 
Feb 04 20:49:43 -- Removing /var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_TensorCompare.cu.o 
Feb 04 20:49:43 /usr/bin/cmake -E remove /var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_TensorCompare.cu.o 
Feb 04 20:49:43 CMake Error at torch_cuda_generated_TensorCompare.cu.o.Debug.cmake:281 (message): 
Feb 04 20:49:43   Error generating file 
Feb 04 20:49:43   /var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_TensorCompare.cu.o 
Feb 04 20:49:43  
Feb 04 20:49:43  
Feb 04 20:49:43 caffe2/CMakeFiles/torch_cuda.dir/build.make:952: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_TensorCompare.cu.o' failed 
Feb 04 20:49:43 make[2]: *** [caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_TensorCompare.cu.o] Error 1 
Feb 04 20:49:43 make[2]: *** Waiting for unfinished jobs.... 
Feb 04 20:49:43 -- Generating temporary cmake readable file: /var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/sparse/cuda/torch_cuda_generated_SparseCUDATensor.cu.o.depend.tmp 
 20:49:43 /usr/bin/cmake -D input_file:FILEPATH=/var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/sparse/cuda/torch_cuda_generated_SparseCUDATensor.cu.o.NVCC-depend -D output_file:FILEPATH=/var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/sparse/cuda/torch_cuda_generated_SparseCUDATensor.cu.o.depend.tmp -D verbose=1 -P /var/lib/jenkins/workspace/cmake/Modules_CUDA_fix/upstream/FindCUDA/make2cmake.cmake 
Feb 04 20:49:44 CMake Warning at /var/lib/jenkins/workspace/cmake/Modules_CUDA_fix/upstream/FindCUDA/make2cmake.cmake:76 (message): 
Feb 04 20:49:44    Removing non-existent dependency file: TH/generic/THStorage.h 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/cuda/ROCmLoops.cuh [hunk 1]
```diff
@@ -253,191 +253,9 @@ static void launch_kernel(int64_t N, const func_t& f, array_t data) {
   AT_CUDA_CHECK(cudaGetLastError());
 }
 
-} // namespace modern
-
-template<typename func_t, int nargs=function_traits<func_t>::arity>
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/ROCmLoops.cuh]: moved to `Loops.cuh`

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 2]
```diff
@@ -281,191 +281,9 @@ static void launch_kernel(int64_t N, const func_t& f, array_t data) {
   AT_CUDA_CHECK(cudaGetLastError());
 }
 
-} // namespace modern
-
-template<typename func_t, int nargs=function_traits<func_t>::arity>
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh]: moved to Loops.cuh

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 3]
```diff
@@ -9,3 +47,192 @@
 #else
 #include <ATen/native/cuda/ROCmLoops.cuh>
 #endif
+
+namespace at { namespace native {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh]: Moved from `CUDALoops.cuh` and `ROCmLoops.cuh`, this part of the code is identical for CUDA and ROCm.

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 4]
```diff
@@ -1,3 +1,41 @@
+
+#pragma once
+
+#include <ATen/detail/FunctionTraits.h>
+
+namespace at { namespace native { namespace modern { namespace detail {
+
+template<typename func_t, int remaining=function_traits<func_t>::arity-1>
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh]: this part is newly added

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 5]
```diff
@@ -9,3 +47,192 @@
...
+      strides[i] = inner_strides[i];
+    }
+
+    if (needs_dynamic_casting<func_t>::check(iter)) {
+      legacy::launch_kernel<launch_size_1d, 1>(numel, [=]GPU_LAMBDA(int idx) {
+        void* out = data[0] + strides[0] * idx;
+        arg0_t result = legacy::invoke(f, &data.data[1], &strides.data[1], &dtypes.data[1], idx);
+        c10::cast_and_store<arg0_t>(dtypes[0], out, result);
+      });
+    } else if (iter.has_contiguous_first_dim() && modern::detail::has_same_arg_types<func_t>::value) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh]: This is the only line changed for this copy-pasted chunk of code.

**Comment by vadimkantorov** [comment]
I hope this enables https://github.com/pytorch/pytorch/issues/9190

**Comment by zasdfgbnm** [comment]
@vadimkantorov Unfortunately this doesn't...

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit cdc04c34:


*Commit cdc04c34 was recently pushed. Waiting for builds...*

---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 6]
```diff
@@ -1,3 +1,41 @@
...
+  static constexpr bool value = true;
+};
+
+template<typename func_t>
+struct has_same_arg_types<func_t, -1> {
+  static constexpr bool value = true;
+};
+
+// simple compile time test for has_same_arg_types:
+using func1_t = int (*)(float, float);
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Loops.cuh]: Yes, this is a compile-time unit test for `has_same_arg_types`. Maybe I should remove it from Loops.cuh and move to somewhere else?

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 7]
```diff
@@ -1,3 +1,41 @@
...
+#include <ATen/detail/FunctionTraits.h>
+
+namespace at { namespace native { namespace modern { namespace detail {
+
+template<typename func_t, int remaining=function_traits<func_t>::arity-1>
+struct has_same_arg_types {
+  using traits = function_traits<func_t>;
+  static constexpr bool value = std::is_same<
+      typename traits::template arg<remaining>::type,
+      typename traits::template arg<remaining-1>::type
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Loops.cuh]: It is specialized as `true` as in the code below. For nullary function, `arity == 0`, therefore `has_same_arg_types<func_t>` will becomes `has_same_arg_types<func_t, function_traits<func_t>::arity-1>` which is `has_same_arg_types<func_t, -1>`

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 8]
```diff
@@ -9,3 +47,195 @@
 #else
 #include <ATen/native/cuda/ROCmLoops.cuh>
 #endif
+
+namespace at { namespace native {
+
+// `needs_dynamic_casting` compares the types expected by iterator
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh]: @ngimel Docs added here

Comment (ngimel) [reply aten/src/ATen/native/cuda/Loops.cuh]: Cool, so it'll be good to go once you move out tests from Loops.cuh.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Loops.cuh]: Test moved to cuda_vectorized_test.cu

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**